modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import memoize from '../../../src/main/resources/META-INF/resources/liferay/util/memoize';
+
+const mockFn = jest.fn((object, number, array, string) => [
+	object,
+	number,
+	array,
+	string,
+]);
+
+const memoizedMockFn = memoize(mockFn);
+
+describe('memoize', () => {
+	it('invokes the provided function with the provided arguments', () => {
+		expect(memoizedMockFn({3: 4}, 4, ['foo', 'bar'], 'baz')).toEqual([
+			{3: 4},
+			4,
+			['foo', 'bar'],
+			'baz',
+		]);
+	});
+
+	it('calls the provided function once, even if the memo is invoked twice', () => {
+		memoizedMockFn({3: 4}, 4, ['foo', 'bar'], 'baz');
+		memoizedMockFn({3: 4}, 4, ['foo', 'bar'], 'baz');
+
+		expect(mockFn).toHaveBeenCalledTimes(1);
+	});
+
+	it('returns a cached value if a function is invoked with same inputs', () => {
+		const mockObject = {
+			12: 24,
+			array: [1, 2, '3'],
+			foo: 'bar',
+		};
+
+		const mockObjectTwo = {
+			array: [4, '5', 6],
+			bar: 'foo',
+			memo: 'ize',
+		};
+
+		const mockFn = jest.fn((object) => object);
+
+		const memoizedMockFn = memoize(mockFn);
+
+		memoizedMockFn(mockObject);
+		memoizedMockFn(mockObjectTwo);
+		memoizedMockFn(mockObject);
+
+		expect(memoizedMockFn(mockObject)).toEqual(memoizedMockFn(mockObject));
+		expect(memoizedMockFn(mockObject)).not.toEqual(
```

Comment:
I think this would be better written as

```js
// This ensures the memo works
expect(memoizedMockFn(mockObject)).toEqual(memoizedMockFn(mockObject));

// This ensures that even though the keys and values in the object are the same, the object itself is not the same reference.
expect(memoizedMockFn(mockObject)).not.toEqual(mockFn(mockObject));
```

Reply:
@bryceosterhaus I think [toBe](https://jestjs.io/docs/expect#tobevalue) is the better avenue for testing object reference/value here, so I'll go with that.

Reply:
❌ `expect(memoizedMockFn(mockObject)).not.toEqual(mockFn(mockObject));` doesn't pass

@bryceosterhaus  It seems that using the `mockObject` variable maintains the reference to the same object, should I use the approach below?

✅  below comparison does pass
```js
expect(
	memoizedMockFn({
		12: 24,
		array: [1, 2, '3'],
		foo: 'bar',
	})
).not.toBe(
	mockFn({
		12: 24,
		array: [1, 2, '3'],
		foo: 'bar',
	})
);
```

Reply:
I also can't figure out what is the reason for this test, like, what should I put in the `it('?', () => ...)`

I wanted this test to test that it's using a cached value. So I can't find a connection between your suggestion and that goal.

Reply:
>  It seems that using the mockObject variable maintains the reference to the same object

Oh yeah that is true, so use the example you provided.

> I wanted this test to test that it's using a cached value.

I think this would still be accurate because you are testing that the value is returning the cached value in this tests
```js
expect(memoizedMockFn(mockObject)).toEqual(memoizedMockFn(mockObject));
```

But then you are also testing the inverse and making sure that the memo'd version doesn't return the same as the non-memo version. This could technically be written as it's own unit test, but I think it works here too